compute_tools/src/compute.rs

```diff
@@ -634,6 +634,48 @@ impl ComputeNode {
         // Update pg_hba.conf received with basebackup.
         update_pg_hba(pgdata_path)?;
 
+        // Place pg_dynshmem under /dev/shm. This allows us to use
+        // 'dynamic_shared_memory_type = mmap' so that the files are placed in
+        // /dev/shm, similar to how 'dynamic_shared_memory_type = posix' works.
+        //
+        // Why on earth don't we just stick to the 'posix' default, you migh
+        // ask.  It turns out that making large allocations with 'posix' doesn't
+        // work very well with autoscaling. The behavior we want is that:
+        //
+        // 1. You can make large DSM allocations, larger than the current RAM
+        //    size of the VM, without errors
+        //
+        // 2. If the allocated memory is really used, the VM is scaled up
+        //    automatically to accommodate that
+        //
+        // We try to make that possible by having swap in the VM. But with the
+        // default 'posix' DSM implementation, we fail step 1, even when there's
+        // plenty of swap available. PostgreSQL uses posix_fallocate() to create
+        // the shmem segment, which is really just a file in /dev/shm in Linux,
+        // but posix_fallocate() on tmpfs returns ENOMEM if the size is larger
+        // than available RAM.
+        //
+        // Using 'dynamic_shared_memory_type = mmap' works around that, because
+        // the Postgres 'mmap' DSM implementation doesn't use
+        // posix_fallocate(). Instead, it uses repeated calls to write(2) to
+        // fill the file with zeros. It's weird that that differs between
```

Comment:
Does this even allocate memory at all? I thought the kernel had a special "zero page" it uses for this kind of thing - but idk if it's smart enough to use that with mmap.

Reply:
You'll need to define "memory" and "allocate" ;-). It does create a file with the right size under /dev/shm. The OS does ensure that there is enough RAM + swap to accommodate it, just like on a regular filesystem, it ensures that there is enough space on the disk. In other words, if you extend a file by writing zeros and it succeeds, and later overwrite the zeros with real data, you will not get ENOSPC. (Not necessarily true with more exotic filesystems with CoW, snapshots etc.).

In addition to that, /dev/shm has a maximum size that you can set when you mount it. You can set that arbitrarily high, higher than the combined RAM + swap space, you'll stll get ENOSPC (or get killed by OOM killer) if you try to create a file larger than what fits in RAM+swap.

The kernel and filesystems (at least regular ones, not sure about tmpfs) have special cases for zerors, but they still do the accounting to make sure there's enough space/memory available, if you overwrite the zeros with something else.

Reply:
But yeah, that's more explicit with posix_fallocate(). The man page specifically says:

> After a successful call to posix_fallocate(), subsequent writes to bytes in the specified range are guaranteed not to fail because of lack of disk space

In practice that is true when write(2) the file with zeros, too, but I guess there's no guarantee.